PSReadLine/StringBuilderTextObjectExtensions.cs

```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.Text;
+
+namespace Microsoft.PowerShell
+{
+    internal static partial class StringBuilderTextObjectExtensions
+    {
+        /// <summary>
+        /// Returns the position of the beginning of the current word as delimited by white space and delimiters
+        /// This method differs from <see cref="ViFindPreviousWordPoint(string)" />:
+        /// When the cursor location is on the first character of a word, <see cref="ViFindPreviousWordPoint(string)" />
+        /// returns the position of the previous word, whereas this method returns the cursor location.
+        ///
+        /// When the cursor location is in a word, both methods return the same result.
+        ///
+        /// This method supports VI "iw" text object.
+        /// </summary>
+        public static int ViFindBeginningOfWordObjectBoundary(this StringBuilder buffer, int position, string wordDelimiters)
+        {
+            // cursor may be past the end of the buffer when calling this method
+            // this may happen if the cursor is at the beginning of a new line
+
+            var i = Math.Min(position, buffer.Length - 1);
+
+            // if starting on a word consider a text object as a sequence of characters excluding the delimiters
+            // otherwise, consider a word as a sequence of delimiters
+
+            var delimiters = wordDelimiters + '\n';

```

Comment:
Given that `buffer.InWord` excludes white characters, it seems adding `\n` to the delimiters here is useless -- imagine `buffer[i]` is `\n`.

Reply:
For the purpose of this method, the newline `\n` character is considered an additional delimiter from the list of default word delimiters.

In any case, you are correct, it has no effect on the `buffer.InWord` test.
However, is absolutely matters when testing whether the cursor is currently on a delimiter.

I tried to clarify intent by rewriting the code slightly.

https://github.com/springcomp/PSReadLine/blob/68de1c6432dcf8e5e25738c595c0c1ad391c4ef7/PSReadLine/StringBuilderTextObjectExtensions.cs#L25-L45

Please, let me know if that clears things up.